### DISCUSSION TITLE
v4 Bootstrapper with .Net 6.0 as a PreReq and WPF UI

### DISCUSSION THREAD

**Discussion by Steve0212a** [wixtoolset/issues#7261] [Questions] [answered]
I am in need of some assistance on how to do what the title says.  I have been trying to read the forum and documentation, but have not found a complete example of how to do this.  I have tried to piece it together.  I did look at https://github.com/wixtoolset/wix4/tree/develop/src/test/burn/TestBA for the UI, but it did not resolve my issue.

I have two problems:
1. Every time I run my exe, it tries to install .Net 6 again even though it is installed (and the detect condition says true).
2. The UI does not appear and it exits with a error

In the below files and attached logs, I have changed some info to "xxxxx" to conceal my client and the product...

Here is my Bundle (notice, I even tried to use the TestBA UI DLL from the wix source repo)
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Version="1.0.0.0" Name="$(env.BUILD_DEFINITIONNAME)"
          Manufacturer="xxxxx" 
          UpgradeCode="xxxxx">
    <Variable Name="REMOVEDATAFILES" Type="formatted" bal:Overridable="yes" Value="" />
    <BootstrapperApplication>
      <Payload SourceFile="xxxxx\WixToolset.Mba.Core.dll" />
      <Payload SourceFile="xxxxx\bin\$(var.Configuration)\net6.0-windows\WixToolset.Mba.Core.dll" />
      <Payload SourceFile="xxxxx\bin\$(var.Configuration)\net6.0-windows\xxxxx.dll" bal:BAFactoryAssembly="yes" />
      <!-- I even tried using the TestBA generated DLL -->
      <!--<Payload SourceFile="C:\Users\xxxxx\Desktop\TestBA\bin\Debug\net6.0-windows\win-x86\TestBA.dll" bal:BAFactoryAssembly="yes" />-->
      <bal:WixManagedBootstrapperApplicationHost AlwaysInstallPrereqs="no" />
    </BootstrapperApplication>

    <Chain>
      <!--Install .Net 6-->
      <PackageGroupRef Id="Net60" />
      <MsiPackage SourceFile="<my MSI>">
        <MsiProperty Name="REMOVEDATAFILES" Value="[REMOVEDATAFILES]" />
      </MsiPackage>
    </Chain>
  </Bundle>
</Wix>
```

And here is my .Net 6 installer
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:netfx="http://wixtoolset.org/schemas/v4/wxs/netfx" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Fragment>
    <netfx:DotNetCoreSearch
        RuntimeType="desktop"
        Platform="x64"
        MajorVersion="6"
        Variable="DOT_NET_VER"
      />

    <PackageGroup Id="Net60">
      <!-- Install .NET 6.0 -->
      <ExePackage Id="Netfx60"
                  DisplayName="Microsoft .NET 6.0 Desktop Runtime"
                  PerMachine="yes" Permanent="yes" Vital="yes" InstallArguments="/norestart /quiet"
                  DetectCondition="DOT_NET_VER &gt;= v6.0.14"
                  bal:PrereqPackage="yes">
        <ExePackagePayload
            Name="windowsdesktop-runtime-6.0.14-win-x64.exe"
            DownloadUrl="https://download.visualstudio.microsoft.com/download/pr/035efed3-6386-4e1d-bcbc-384a20ebf47e/abfbea2303e0ce9cb15d430314e5858f/windowsdesktop-runtime-6.0.14-win-x64.exe"
            Hash="5281E817571AFA6C5D848ED9A3FEC4CB568AD8C05633CBD97E65E52F4A929D9947C390B9B3A72DA6464C89450576D318B253D190776BD357327FC27031C7ACDF"
            Size="57339048"
          />
        <ExitCode Value="0" Behavior="success"/>
        <ExitCode Behavior="scheduleReboot"/>
      </ExePackage>
    </PackageGroup>
  </Fragment>
</Wix>
```

Here is WixToolset.Mba.Host.config (found out that the config file must be named this as I got an error when it had a different name)
```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <configSections>
    <sectionGroup name="wix.bootstrapper" type="WixToolset.Mba.Host.BootstrapperSectionGroup, WixToolset.Mba.Host">
      <section name="host" type="WixToolset.Mba.Host.HostSection, WixToolset.Mba.Host" />
    </sectionGroup>
  </configSections>
  <startup>
    <supportedRuntime version="v6.0" />
  </startup>
  <wix.bootstrapper>
    <host assemblyName="UsbShutdownInstallerWpfUi" />
  </wix.bootstrapper>
</configuration>
```

Here is the output log:
[log.txt](https://github.com/wixtoolset/issues/files/10913440/log.txt)

I would think the errors in the log that are important for why the UI is not loading are these:
```
[25B4:2428][2023-03-07T14:44:43]i006: Bootstrapper application requested to be reloaded.
[25B4:2428][2023-03-07T14:44:43]e000: Error 0x80131700: Failed to get the CLR runtime info using the application configuration file path.
[25B4:2428][2023-03-07T14:44:43]e000: Error 0x81f403ea: The prerequisites were already installed. The bootstrapper application will not be reloaded to prevent an infinite loop.
[25B4:2428][2023-03-07T14:44:43]i000: Loading prerequisite bootstrapper application because managed host could not be loaded, error: 0x81f403ea.
```

So:
1. Why is the prereq loading every time?  In the log, it says it is present, yet I get the popup to install prereqs even though I have AlwaysInstallPrereqs="no".  Is it because the UI is failing to load?
```
[25B4:2428][2023-03-07T14:44:30]i052: Condition 'DOT_NET_VER >= v6.0.14' evaluates to true.
[25B4:2428][2023-03-07T14:44:30]i101: Detected package: Netfx60, state: Present, cached: Yes, install registration state: (permanent), cache registration state: (permanent)
```

2. Why am I getting "Failed to get the CLR runtime info"?  Is there a different way to specify it for .Net 6 in the WixToolset.Mba.Core.dll than supportedRuntime?

3. I am very new to Wix v4.  Anything else you see that I did wrong?

Thanks for any help.

**Comment by rseanhall** [comment, marked as answer]
If your managed BA is built with .NET 6 then you need to use `bal:WixDotNetCoreBootstrapperApplicationHost`. `bal:WixManagedBootstrapperApplicationHost` is only for .NET Framework managed BAs. There should probably be a warning or error if `bal:BAFactoryAssembly="yes"` is specified with `bal:WixManagedBootstrapperApplicationHost` so you should create an issue for that if you think that would've helped you here.

**Comment by Steve0212a** [comment]
I had tried that previously, but was unsure if that was correct.

I updated it to this but it still fails
`<bal:WixDotNetCoreBootstrapperApplicationHost AlwaysInstallPrereqs="no" />`

Here are the errors in the log:
```
[1DD0:0EC0][2023-03-07T16:19:30]i006: Bootstrapper application requested to be reloaded.
[1DD0:0EC0][2023-03-07T16:19:30]e000: Error 0x80008083: GetHostfxrPath failed
[1DD0:0EC0][2023-03-07T16:19:30]e000: Error 0x80008083: Failed to find hostfxr.
[1DD0:0EC0][2023-03-07T16:19:30]e000: Error 0x81f403ea: The prerequisites were already installed. The bootstrapper application will not be reloaded to prevent an infinite loop.
[1DD0:0EC0][2023-03-07T16:19:30]i000: Loading prerequisite bootstrapper application because .NET Core host could not be loaded, error: 0x81f403ea.
```

**Reply by rseanhall** [reply]
hostfxr error codes are documented at https://github.com/dotnet/runtime/blob/main/src/native/corehost/error_codes.h. You appear to be installing x64 .NET but you appear to have an x86 BA.

**Reply by Steve0212a** [reply]
making progress.  You were correct about x86 vs x64.  I have that all cleaned up now (everything is x64).

Now, when I run my EXE, it just starts and exits.   This is from the log.  0x80131524 seems to be DllNotFoundException.  If that is the case, which dll am I missing?

Log:
```
[2748:0418][2023-03-07T17:21:10]i000: Loading .NET Core FDD bootstrapper application.
[2748:0418][2023-03-07T17:21:10]e000: Error 0x80131524: Failed to create the .NET Core bootstrapper application.
[2748:0418][2023-03-07T17:21:10]e000: Error 0x80131524: Failed to create BA.
[2748:0418][2023-03-07T17:21:10]e000: Error 0x80131524: Failed to load BA.
[2748:0418][2023-03-07T17:21:10]e000: Error 0x80131524: Failed while running 
```

This is my updated bundle:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Version="1.0.0.0" Name="$(env.BUILD_DEFINITIONNAME)"
          Manufacturer="xxxxx" 
          UpgradeCode="xxxxx">
    <Variable Name="REMOVEDATAFILES" Type="formatted" bal:Overridable="yes" Value="" />
    <BootstrapperApplication>
      <Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\xxxxxWpfUi.deps.json" />
      <Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\xxxxxWpfUi.runtimeconfig.json" />
      <Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\WixToolset.Mba.Host.config" />
      <Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\WixToolset.Mba.Core.dll" />
      <Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\xxxxxWpfUi.dll" bal:BAFactoryAssembly="yes" />
      <bal:WixDotNetCoreBootstrapperApplicationHost AlwaysInstallPrereqs="no" />
    </BootstrapperApplication>

    <Chain>
      <!--Install .Net 6-->
      <PackageGroupRef Id="Net60" />
      <MsiPackage SourceFile="..\xxxxxMsi\bin\x64\$(var.Configuration)\xxxxxMsi.msi">
        <MsiProperty Name="REMOVEDATAFILES" Value="[REMOVEDATAFILES]" />
      </MsiPackage>
    </Chain>
  </Bundle>
</Wix>
```

My UI Dll has no references to be missing (other than WIX):
<img width="276" alt="image" src="https://user-images.githubusercontent.com/18215939/223569005-05897316-7091-49f2-85de-beb7c78f6294.png">

One quick question - is my UI supposed to be a DLL or an EXE?

Any other ideas?

Thanks for the help - appreciate it.

**Reply by rseanhall** [reply]
You should be able to debug it in Visual Studio since it has loaded the .NET runtime at that point. But it's probably because `WixToolset.Mba.Core.dll` requires `mbanative.dll`.

**Reply by Steve0212a** [reply]
That did it - Thanks for the help!

Added to my bundle:
`<Payload SourceFile="xxxxxWpfUi\bin\x64\$(var.Configuration)\net6.0-windows\runtimes\win-x64\native\mbanative.dll" />`

**Reply by Shivapatne** [reply]
@Steve0212a Can the BA UI launch on a fresh machine without manually installing .NET Core?

**Reply by Steve0212a** [reply]
> @Steve0212a Can the BA UI launch on a fresh machine without manually installing .NET Core?

you should be able to do a self contained BA UI.  I would post that as its own question in Discussions
